File: examples/Cpp/AzureKinectViewer.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include <k4a/k4a.h>

#include "assert.h"

#include <math.h>
#include <atomic>
#include <csignal>
#include <ctime>
#include <iostream>

#include "Open3D/Open3D.h"

using namespace open3d;

void PrintUsage() {
    PrintOpen3DVersion();
    // clang-format off
    utility::LogInfo("Options: \n");
    utility::LogInfo("--config  Config .json file (default: none)\n");
    utility::LogInfo("--list    List the currently connected K4A devices\n");
    utility::LogInfo("--device  Specify the device index to use (default: 0)\n");
    utility::LogInfo("-a        Align depth with color image (default: disabled)\n");
    utility::LogInfo("-h        Print this helper\n");
    // clang-format on
}

int main(int argc, char **argv) {
    // Parse arguments
    if (utility::ProgramOptionExists(argc, argv, "-h")) {
        PrintUsage();
        return 0;
    }

    if (utility::ProgramOptionExists(argc, argv, "--list")) {
        io::AzureKinectSensor::ListDevices();
        return 0;
    }

    io::AzureKinectSensorConfig sensor_config;
    if (utility::ProgramOptionExists(argc, argv, "--config")) {
        auto config_filename =
                utility::GetProgramOptionAsString(argc, argv, "--config", "");
        if (!io::ReadIJsonConvertibleFromJSON(config_filename, sensor_config)) {
            utility::LogInfo("Invalid sensor config\n");
            return 1;
        }
    } else {
        utility::LogInfo("Use default sensor config\n");
    }

    int sensor_index =
            utility::GetProgramOptionAsInt(argc, argv, "--device", 0);
    if (sensor_index < 0 || sensor_index > 255) {
        utility::LogError("Sensor index must between [0, 255]: {}\n",
                          sensor_index);
        return 1;
    }

    bool enable_align_depth_to_color =
            utility::ProgramOptionExists(argc, argv, "-a");

    // Init sensor
    io::AzureKinectSensor sensor(sensor_config);
    if (!sensor.Connect(sensor_index)) {
        utility::LogError("Failed to connect to sensor, abort.\n");
        return 1;
    }

    // Start viewing
    bool flag_exit = false;
    bool is_geometry_added = false;
    visualization::VisualizerWithKeyCallback vis;
    vis.RegisterKeyCallback(GLFW_KEY_ESCAPE,
                            [&](visualization::Visualizer *vis) {
                                flag_exit = true;
                                return false;
                            });

    vis.CreateVisualizerWindow("Open3D Azure Kinect Recorder", 1920, 540);
    do {
        auto im_rgbd = sensor.CaptureFrame(enable_align_depth_to_color);
        if (im_rgbd == nullptr) {
            utility::LogInfo("Invalid capture, skipping this frame\n");
            continue;
        }

        if (!is_geometry_added) {
            vis.AddGeometry(im_rgbd);
            is_geometry_added = true;
        }

        // Update visualizer
        vis.UpdateGeometry();
        vis.PollEvents();
        vis.UpdateRender();

    } while (!flag_exit);

    return 0;
}

File: examples/Python/ReconstructionSystem/sensors/azure_kinect_viewer.py
import argparse
import datetime
import open3d as o3d
import os


kbd_escape = 256
flag_stop = False
def escape_callback(vis):
    global flag_stop, recorder
    flag_stop = True
    return False


def main(sensor, align_depth_to_color):
    vis = o3d.visualization.VisualizerWithKeyCallback()
    vis.register_key_callback(kbd_escape, escape_callback)

    vis_geometry_added = False
    vis.create_window('viewer', 1920, 540)

    while not flag_stop:
        rgbd = sensor.capture_frame(align_depth_to_color)
        if rgbd is None:
            continue

        if not vis_geometry_added:
            vis.add_geometry(rgbd)
            vis_geometry_added = True

        vis.update_geometry()
        vis.poll_events()
        vis.update_renderer()


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Azure kinect mkv recorder.')
    parser.add_argument('--config', type=str,
                        help='input json kinect config')
    parser.add_argument('--list', action='store_true',
                        help='list available azure kinect sensors')
    parser.add_argument('--device', type=int, default=0,
                        help='input kinect device id')
    parser.add_argument('-a', '--align_depth_to_color', action='store_true',
                        help='enable align depth image to color')

    args = parser.parse_args()

    if args.list:
        o3d.io.AzureKinectSensor.list_devices()
        exit()

    if args.config is not None:
        config = o3d.io.read_azure_kinect_sensor_config(args.config)
    else:
        config = o3d.io.AzureKinectSensorConfig()

    device = args.device
    if device < 0 or device > 255:
        print('Unsupported device id, fall back to 0')
        device = 0

    azure_kinect_sensor = o3d.io.AzureKinectSensor(config)
    rc = azure_kinect_sensor.connect(device)
    if rc != 0:
        print('Failed to connect to sensor')
        exit()

    main(azure_kinect_sensor, args.align_depth_to_color)